jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContentProducer.java

```diff
@@ -0,0 +1,259 @@
+//
+
+package org.eclipse.jetty.server;
+
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jetty.http.BadMessageException;
+import org.eclipse.jetty.http.HttpStatus;
+
+class AsyncContentProducer implements ContentProducer
+{
+    private final HttpChannel _httpChannel;
+    private HttpInput.Interceptor _interceptor;
+    private HttpInput.Content _rawContent;
+    private HttpInput.Content _transformedContent;
+    private boolean _error;
+
+    private long _firstByteTimeStamp = Long.MIN_VALUE;
+    private long _rawContentArrived;
+
+    public AsyncContentProducer(HttpChannel httpChannel)
+    {
+        _httpChannel = httpChannel;
+    }
+
+    @Override
+    public void recycle()
+    {
+        _interceptor = null;
+        _rawContent = null;
+        _transformedContent = null;
+        _error = false;
+        _firstByteTimeStamp = Long.MIN_VALUE;
+        _rawContentArrived = 0L;
+    }
+
+    @Override
+    public HttpInput.Interceptor getInterceptor()
+    {
+        return _interceptor;
+    }
+
+    @Override
+    public void setInterceptor(HttpInput.Interceptor interceptor)
+    {
+        this._interceptor = interceptor;
+    }
+
+    @Override
+    public int available()
+    {
+        HttpInput.Content content = nextTransformedContent();
+        return content == null ? 0 : content.remaining();
+    }
+
+    @Override
+    public boolean hasContent()
+    {
+        return _rawContent != null;
+    }
+
+    @Override
+    public boolean isError()
+    {
+        return _error;
+    }
+
+    @Override
+    public void checkMinDataRate()
+    {
+        long minRequestDataRate = _httpChannel.getHttpConfiguration().getMinRequestDataRate();
+        if (minRequestDataRate > 0 && _firstByteTimeStamp != Long.MIN_VALUE)
+        {
+            long period = System.nanoTime() - _firstByteTimeStamp;
+            if (period > 0)
+            {
+                long minimumData = minRequestDataRate * TimeUnit.NANOSECONDS.toMillis(period) / TimeUnit.SECONDS.toMillis(1);
+                if (getRawContentArrived() < minimumData)
+                {
+                    BadMessageException bad = new BadMessageException(HttpStatus.REQUEST_TIMEOUT_408,
+                        String.format("Request content data rate < %d B/s", minRequestDataRate));
+                    if (_httpChannel.getState().isResponseCommitted())
+                        _httpChannel.abort(bad);
+                    failCurrentContent(bad);
+                    throw bad;
+                }
+            }
+        }
+    }
+
+    @Override
+    public long getRawContentArrived()
+    {
+        return _rawContentArrived;
+    }
+
+    @Override
+    public boolean consumeAll(Throwable x)
+    {
+        failCurrentContent(x);
+        // A specific HttpChannel mechanism must be used as the normal produceContent()
+        // mechanism does not guarantee that all available content will be returned;
+        // tests in StreamResetTest cover this.
```

Comment:
This comment is pretty poor (it's definitely a rough brain dump that needs to be cleared). What it means is that with the following loop:

```java
while (_httpChannel.needContent())
{
    HttpInput.Content content = _httpChannel.produceContent();
    ...
}
```

you have no guarantee that even if the channel does have some content immediately available that it won't deliver it asynchronously by calling `HttpInput.wakeup()` when its ready to be read. So far, this is only true for H2 because calling `Stream.demand(1)` does not guarantee that `onData()` will be called before the demand call returns, even if the stream does contain some content.